design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,1146 @@
+        crossplaneVersion: "2.1.0"
+        # Manifests to apply as part of the test.
+        manifests:
+          - apiVersion: platform.example.com/v1alpha1
+            kind: Cluster
+            metadata:
+              name: test-cluster
+            spec:
+              version: 1.33
+              region: us-west1
+        # Extra resources that should be installed in the cluster before the test is
+        # executed. This allows for configuration of provider credentials, for example.
+        extraResources:
+          - apiVersion: gcp.upbound.io/v1beta1
+            kind: ProviderConfig
+            metadata:
+              name: default
+            spec:
+              credentials:
+                secretRef:
+                  key: credentials
+                  name: gcp-credentials
+                  namespace: crossplane-system
+                source: Secret
+              projectID: example-dot-com-testing
+          - apiVersion: v1
+            data:
+              credentials: c3VwZXIgc2VjcmV0IHBhc3N3b3JkIGluc2lkZQo=
+            kind: Secret
+            metadata:
+              name: gcp-credentials
+              namespace: crossplane-system
+        # Conditions the test will wait for the applied resources to have.
+        defaultConditions:
+          - Ready
+        # Whether to skip deletion of applied resources.
+        skipDelete: false
+        # Timeout for the test.
+        timeoutSeconds: 300
+        # Timeout for post-test cleanup, which tries to ensure no resources are left behind.
+        cleanupTimeoutSeconds: 600
+```
+
+The tooling can either create a local, ephemeral test cluster (using `kind`) in
+which to run e2e tests, or run them against an arbitrary kubeconfig
+context. Either way, the test is converted into an `uptest` test case and
+executed against the test cluster. The tooling takes care of cleaning up
+resources after the test runs, to try and avoid potentially leaving behind any
+cloud resources that were created.
+
+#### Extensibility
+
+Composition tests have two phases: render and assertion. The render phase is a
+core part of the tooling, but assertion could be open to extension. The API
+above includes a `type` field for assertions, allowing for other assertion
+frameworks to be added. A new type could be introduced that runs an arbitrary
+command and provides the results of the render on standard input, allowing for
+assertions to be written using any tool the user prefers.
+
+E2E tests are less extensible, since they are executed using uptest. Given the
+comparatively higher complexity of e2e tests (which deal with actual clusters
+and potentially real cloud resources), it is likely more appropriate to
+introduce extensibility points in uptest rather than the wrapper provided by the
+DevEx tooling.
+
+## Appendix: Proposed New Commands
+
+For clarity, this is the full tree of commands this design proposes adding to
+the `crossplane` CLI:
+
+* `crossplane project`
```

Comment:
Are `generate`, `test` etc fundamentally coupled to a project? You couldn't test something outside a project could you? If so - should everything be under `crossplane project`? `crossplane project generate`, etc?

Reply:
actually the goal is to be able to use `crossplane test` directly, without a project. I guess there could be also a `crossplane project test` as a thin wrapper of `crossplane test` for project-managed configuration